### PR TITLE
Add `.apply()` operator to stages

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Introduced pipe operator to allow more customizable Pipeline code, i.e. extracting common logic.

Currently it's a draft PR, just to show concept

### CODE REVIEW

**Comment by cangencer** [comment]
verify

**Comment by cangencer** [comment]
I like the idea overall, it's similar to how an extension method would work. Still have to think a little about the naming of the method, if it should be called `pipe` or something else.

**Comment by TomaszGaweda** [comment]
@cangencer Some time ago I read, that newest Java version provides String.transform method with similar behaviour and they plan to apply it to other types as well.
Pros for this name: consistency. Cons: Possible name clash.
Currently I've used the name of Pipe operator from Scala FP libraries. I'm not only sure if users' won't be confused why Pipeline has method pipe :P

**Comment by cangencer** [comment]
I  was thinking of calling it `apply` perhaps, in that you're applying a function to the pipeline

**Comment by cangencer** [comment]
@TomaszGaweda 

```
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java:479: Line is longer than 123 characters (found 136). [LineLength]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStageWithKey.java:29: Using the '.*' form of import should be avoided - com.hazelcast.jet.function.*. [AvoidStarImport]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStageWithKey.java:413:5: Missing a Javadoc comment. [JavadocMethod]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStageWithKey.java:414: Line is longer than 123 characters (found 146). [LineLength]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/StreamStage.java:189:5: Missing a Javadoc comment. [JavadocMethod]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/StreamStageWithKey.java:25: Using the '.*' form of import should be avoided - com.hazelcast.jet.function.*. [AvoidStarImport]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/StreamStageWithKey.java:107:5: Missing a Javadoc comment. [JavadocMethod]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/StreamStageWithKey.java:108: Line is longer than 123 characters (found 149). [LineLength]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java:34: Using the '.*' form of import should be avoided - java.util.*. [AvoidStarImport]
09:32:42 [ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java:116:5: Static variable definition in wrong order. [DeclarationOrder]
09:32:42 Audit done.```

**Comment by viliam-durina** [comment]
verify

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java [hunk 1]
```diff
@@ -468,6 +468,51 @@
...
+     * }</pre>
+     *
+     * The {@code addSpecialMapping} method can then be re-used in multiple
+     * pipelines.
+     *
+     * @param transformFn a function to transform this stage to another stage
+     * @param <R> type of the returned stage
+     */
+    @Nonnull
+    default <R extends Stage> R apply(@Nonnull FunctionEx<BatchStage<T>, ? extends R> transformFn) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java]: It should be `? super T`

**Comment by cangencer** [comment]
verify

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java (hunk 1 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java]: I ended up adding `? super BatchStage`. This way the actual parameter can be for example `FunctionEx<BatchStage<? extends Number>, Whatever>`. Without it, the function input has to be `BatchStage<Integer>`.

I will not pretend that I understand why, @mtopolnik might be able to explain it :).

**Comment by viliam-durina** [comment]
Fixes #1310

**Comment by viliam-durina** [comment]
verify

**Comment by cangencer** [comment]
verify

**Comment by mtopolnik** [comment]
verify

**Comment by cangencer** [comment]
@TomaszGaweda thanks for the contribution!

**Comment by TomaszGaweda** [comment]
@cangencer Thank you all for your help :) I did only the first part ;)

**Pull request merged**